### ISSUE TITLE
Wrong ID number for tilesets based on tilesets image

### ISSUE THREAD

**Issue by aornano** [mfessenden/SKTiled#30]
I've a map maked with "Tiled Map Editor" version 1.4.2 and all is perfect using this editor.
When I try to load all map and assets inside my game, strange things happened during the loading of tiles taken by tilesets based on tilesets image (multiple tiles in one big image). 
It seems the ID taken is the previous in the tile sequence, so instead of showing TILE ID 4, SKTiled show ID 3.

I was searching for the exact responsible method and I've found it:

```
public func getLocalID(forGlobalID gid: Int) -> Int {
        // firstGID is greater than 0 only when added to a tilemap
        let id = (firstGID > 0) ? (gid - firstGID) : gid
        // if the id is less than zero, return the gid
        return (id < 0) ? gid : id
    }
```
**if I change this line the tilesets based on tilesets showed correctly all tiles, but the rest of the map is total black**
` let id = (firstGID > 0) ? (gid - firstGID) + 1 : gid`

Any idea about it? Could I distinguish tilesets based on tilesets from traditional tilesets to apply a condition and solve it?

Thank you in advance for all your support.

**Comment by mfessenden** [comment]
Do you have a sample scene you can send me to illustrate the problem? 

The global id (gid) represents the tile data's *internal* id plus the tileset's `firstGID` value. Internally, the tileset stores tile data in an array starting with 0 with the upper bound being the number of tiles.

The value stored in the map file is the global id, so the `SKTileset.getLocalID(forGlobalID :)` method is used to resolve a global id to a local id.

**Comment by aornano** [comment]
This tileset started with ID 0 for the first upper left tile.
As you can see to these screenshots , I've try to put the grass tile but in the map it appear with the previous tile (the grass with the stones).
So it takes the ID 3 instead of ID 4.
![Screenshot 2020-11-05 at 14 39 48](https://user-images.githubusercontent.com/5380705/98248170-dda24a00-1f74-11eb-9959-776c9aaddfba.png)
![Screenshot 2020-11-05 at 14 40 06](https://user-images.githubusercontent.com/5380705/98248176-df6c0d80-1f74-11eb-91d2-3effa717bf30.png)

**Comment by aornano** [comment]
Ok I've found the problem. 

**This issue is not from your code, your code is correct.**

The problem is the default value of TiledMap when you create the TSX file:

![Screenshot 2020-11-06 at 15 27 13](https://user-images.githubusercontent.com/5380705/98377375-04c84c80-2045-11eb-8a71-357fd1c86694.png)

So this is not a bug, this report can be closed. Thank you.

**Comment by mfessenden** [comment]
The problem actually lies within your tileset file. 

I noticed in your second post that the image you were loading was too large for a 3x3 tileset at 128x64px. I changed this...

```xml
<?xml version="1.0" encoding="UTF-8"?>
<tileset version="1.4" tiledversion="1.4.2" name="grass1" tilewidth="128" tileheight="64" margin="1" tilecount="9" columns="3">
 <image source="handpainted.png" width="512" height="256"/>
</tileset>

```

... to this...

```xml
<?xml version="1.0" encoding="UTF-8"?>
<tileset version="1.4" tiledversion="1.4.2" name="grass1" tilewidth="128" tileheight="64" tilecount="16" columns="4">
 <image source="handpainted.png" width="512" height="256"/>
</tileset>
```

...and the preview in **Tiled** now matches what you're seeing in your iOS app:

![Ismoetric Preview][iso-preview-img]


When the `SKTileset` loads in your source spritesheet image, I'm calculating the bounding rect for each tile based on the total tile count, the source image dimensions and spacing/margin values. So in **SKTiled**, the tile with **id 3** is actually this guy:


![Missing Tile][dirt-tile-img]


Bjorn is creating his tileset images a little differently, so you've given me a great use-case for reviewing how we calculate our tile breakup 😉


Technically, this is a bug so I'll keep it closed, but will add it to the list of things to fix.


[iso-preview-img]:https://user-images.githubusercontent.com/911532/98389968-c883f480-2022-11eb-9688-33b23fc1b55f.png
[dirt-tile-img]:https://user-images.githubusercontent.com/911532/98391380-9378a180-2024-11eb-84fd-aba197c37da2.png
